File: Java/Implementations/Strategy/Duck.java
public class Duck {
	private String name;
	private Movement movement;
	private Position position;

	public Duck(String name, Movement movement) {
		this.name = name;
		this.movement = movement;
		this.position = new Position(0, 0);
	}

	public void moveTo(Position dest) {
		System.out.println(this.name +  " " + this.movement.moveTo(this.position, dest) + ".");
		this.position = dest;
	}
}

File: Haskell/Implementations/Strategy/Duck.hs
module Duck
    ( Duck(..)
    , mkDuck
    , moveDuck
    ) where

import Movement (Movement, Position(..))

data Duck = MkDuck
    { name     :: String
    , movement :: Movement
    , pos      :: Position
    }

mkDuck :: String -> Movement -> Duck
mkDuck n m = MkDuck n m (MkPosition 0 0)

moveDuck :: Position -> Duck -> IO Duck
moveDuck dest (MkDuck name moveFn currPos) = do
    putStrLn (name <> " " <> moveFn currPos dest <> ".")
    pure (MkDuck name moveFn dest)
